### ISSUE TITLE
Unnecessary call in Dispatcher

### ISSUE THREAD

**Issue by Crell** [antidot-framework/antidot-event-dispatcher#1]
Hi there.  Yay, more PSR-14!

I spotted a possible issue in your Dispatcher, though:

https://github.com/kpicaza/antidot-event-dispatcher/blob/master/src/EventDispatcher.php#L30

The Provider should be returning the listeners to call directly.  Not a callable that produces a callable.  Viz, the `$callableListener` is what should be called with `$event`.  Line 30 seems to be unnecessary, or at least incompatible.

**Comment by kpicaza** [comment]
Thank you very much for your interest.

I will correct it as soon as possible.

The idea of using the callable instead of using the listener directly, was to delay as much as possible the construction of the listener or avoid it once the propagation of the event was stopped.

**Comment by Crell** [comment]
You can do that, but that should be done in the Provider.  The Provider can return a callable that just delegates to another callable that it lazily creates.  That's fine, and is the common way to pull something out of a Container, for instance.  But what gets to the Dispatcher should be just "a callable I can do `$callable($event)` on".

If you want to lazily create the callables themselves, use `yield` inside the Provider.  It's quite cool. :smile:

**Comment by kpicaza** [comment]
I have opened a pull request https://github.com/kpicaza/antidot-event-dispatcher/pull/2 with a possible solution that still allows the Lazy loading of the listeners from the provider.

Many thanks for the clarification, it has been very helpful

**Comment by kpicaza** [comment]
I rebased the PR that I think solves the problem that you mentioned in the Dispatcher.

I appreciate a lot your help and I would be happy to keep in touch if you found any other possible problem.

Greetings and Yes! Psr Rocks!! ;-D

**Comment by Crell** [comment]
:smile: 

The main question I have with the design here is why you are bringing back event name strings.  PSR-14 was specifically designed to eliminate the need for that, as it's strictly less flexible than just using the event itself.  That is, your provider will only work with a subset of events (those that have your event interface), and will have fewer things it can match on than a generic one.

I don't really see the benefit of that approach...

**Comment by kpicaza** [comment]
Thanks again. I completely agree with you. The reason why I decided to use event name strings was more mechanical than logical: I implemented a pattern that I already knew about instead of investigating a little more.

I consider your point of view, so in the next release, I will apply it correctly.

Thank you very much for your constructive criticism. Your help encourages me to continue learning and improving
